### PR TITLE
Add faulty injection in writer fuzzer

### PR DESCRIPTION
**Pull request by kewang1024**

### CODE REVIEW

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 1]
```diff
@@ -340,7 +343,26 @@ void WriterFuzzer::go() {
...
 
-    auto tempDirPath = exec::test::TempDirectoryPath::create();
+    auto tempDirPath = exec::test::TempDirectoryPath::create(true);
+
+    bool writeErrorInjected = false;
+    auto fs = std::dynamic_pointer_cast<tests::utils::FaultyFileSystem>(
+        filesystems::getFileSystem(tempDirPath->getPath(), {}));
+    if (partitionKeys.empty() && vectorFuzzer_.coinToss(1)) {
+      fs->setFileInjectionError(
+          std::make_exception_ptr(VeloxRuntimeError(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:353]: VELOX_FAIL with specific error should be good enough

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 2]
```diff
@@ -436,7 +462,14 @@ void WriterFuzzer::verifyWriter(
 
   const auto maxDrivers =
       boost::random::uniform_int_distribution<int32_t>(1, 16)(rng_);
-  const auto result = veloxToPrestoResult(execute(plan, maxDrivers));
+  RowVectorPtr result;
+  try {
+    result = veloxToPrestoResult(execute(plan, maxDrivers));
+  } catch (VeloxRuntimeError&) {
+    VELOX_CHECK(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:469]: We can tolerate the error injection

File: velox/dwio/common/FileSink.cpp [hunk 3]
```diff
@@ -145,6 +145,18 @@ LocalFileSink::LocalFileSink(const std::string& name, const Options& options)
   writeFile_ = fs->openFileForWrite(name_);
 }
 
+LocalFileSink::LocalFileSink(
+    const std::string& name,
+    const Options& options,
+    const bool initializeWriter)
```
Comment (xiaoxmeng) [inline velox/dwio/common/FileSink.cpp:151]: drop const for bool

File: velox/dwio/common/FileSink.h [hunk 4]
```diff
@@ -166,9 +166,14 @@ class LocalFileSink : public FileSink {
...
  protected:
+  // Only used by FaultyFileSink to initialize writeFile_ as FaultyWriteFile.
+  LocalFileSink(
+      const std::string& name,
+      const Options& options,
+      const bool initializeWriter);
   void doClose() override;
 
- private:
+ protected:
```
Comment (xiaoxmeng) [inline velox/dwio/common/FileSink.h:176]: Just remove private

File: velox/dwio/common/tests/FaultyFileSink.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#include "velox/dwio/common/FileSink.h"
+#include "velox/dwio/common/MetricsLog.h"
+
+namespace facebook::velox::dwio::common {
+using namespace facebook::velox::io;
+
+class FaultyFileSink : public LocalFileSink {
+ public:
+  FaultyFileSink(
+      const std::string& faultyName,
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/FaultyFileSink.h:36]:
```
s/faultyName/faultyFilePath/
```

File: velox/dwio/common/tests/FaultyFileSink.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      const std::string& name,
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/FaultyFileSink.h:37]: I think you could substr this inside FaultyFileSink ctor

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 7]
```diff
@@ -123,7 +136,9 @@ class WriterFuzzer {
       const std::vector<std::string>& bucketColumns,
       int32_t sortColumnOffset,
       const std::vector<std::shared_ptr<const HiveSortingColumn>>& sortBy,
-      const std::string& outputDirectoryPath);
+      const std::string& outputFaultyDirectoryPath,
+      const std::string& outputDirectoryPath,
+      const bool writeErrorInjected);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:141]:
```
bool writeErrorInjected
```

File: velox/dwio/common/FileSink.h [hunk 8]
```diff
@@ -166,9 +166,13 @@ class LocalFileSink : public FileSink {
   static void registerFactory();
 
  protected:
+  // Only used by FaultyFileSink to initialize writeFile_ as FaultyWriteFile.
```
Comment (xiaoxmeng) [inline velox/dwio/common/FileSink.h:169]:
```
// 'initializeWriter' is false if it is used by FaultyFileSink which setups the write file through the fault filesystem.
```

File: velox/dwio/common/tests/FaultyFileSink.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace facebook::velox::dwio::common {
+using namespace facebook::velox::io;
+
+class FaultyFileSink : public LocalFileSink {
+ public:
+  FaultyFileSink(const std::string& faultyFilePath, const Options& options);
+
+ private:
+  std::string faultyFilePath_;
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/FaultyFileSink.h:38]:
```
const std::string faultyFilePath_;
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 10]
```diff
@@ -340,7 +355,20 @@ void WriterFuzzer::go() {
     }
     auto input = generateInputData(names, types, partitionOffset);
 
-    auto tempDirPath = exec::test::TempDirectoryPath::create();
+    auto tempFaultyDirPath = exec::test::TempDirectoryPath::create(true);
+
+    // Generates file system write error with certain possibility
+    auto faultyFs = std::dynamic_pointer_cast<FaultyFileSystem>(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:361]:
```
    const auto outputDirPath = exec::test::TempDirectoryPath::create(FLAGS_file_system_error_injection);
```

File: velox/exec/tests/utils/TempDirectoryPath.h [hunk 11]
```diff
@@ -45,6 +45,10 @@ class TempDirectoryPath {
     return path_;
   }
 
+  const std::string& getDelegatePath() const {
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/TempDirectoryPath.h:48]:
```
/// The actual file path if fault injection is enabled.
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 12]
```diff
@@ -244,6 +253,12 @@ class WriterFuzzer {
   std::shared_ptr<memory::MemoryPool> writerPool_{
       rootPool_->addAggregateChild("writerFuzzerWriter")};
   VectorFuzzer vectorFuzzer_;
+
+  const std::shared_ptr<FaultyFileSystem> faultyFs_ =
+      std::dynamic_pointer_cast<FaultyFileSystem>(
+          filesystems::getFileSystem("faulty:/tmp", {}));
+  std::atomic<uint64_t> injectedErrorCount_{0};
+  const std::string injectedErrorMsg_ = "Injected Faulty File Error";
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:261]: Put const member first
```
const std::string injectedErrorMsg_{"Injected Faulty File Error"};
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 13]
```diff
@@ -292,6 +307,16 @@ void WriterFuzzer::go() {
   auto startTime = std::chrono::system_clock::now();
   size_t iteration = 0;
 
+  // Faulty fs will generate file system write error with certain possibility
+  if (FLAGS_file_system_error_injection) {
+    faultyFs_->setFileInjectionHook([&](FaultFileOperation* op) {
+      if (vectorFuzzer_.coinToss(1)) {
+        injectedErrorCount_++;
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:314]: nit: ++injectedErrorCount_

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 14]
```diff
@@ -436,7 +463,18 @@ void WriterFuzzer::verifyWriter(
...
       boost::random::uniform_int_distribution<int32_t>(1, 16)(rng_);
-  const auto result = veloxToPrestoResult(execute(plan, maxDrivers));
+  RowVectorPtr result;
+  const uint64_t prevInjectedErrorCount = injectedErrorCount_;
+  try {
+    result = veloxToPrestoResult(execute(plan, maxDrivers));
+  } catch (VeloxRuntimeError& error) {
+    if (injectedErrorCount_ > prevInjectedErrorCount) {
+      VELOX_CHECK(
+          error.message() == injectedErrorMsg_,
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:473]:
```
VELOX_CHECK_EQ(xxx, yyy,
Unexpected writer fuzzer failure
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 15]
```diff
@@ -436,7 +463,18 @@ void WriterFuzzer::verifyWriter(
 
   const auto maxDrivers =
       boost::random::uniform_int_distribution<int32_t>(1, 16)(rng_);
-  const auto result = veloxToPrestoResult(execute(plan, maxDrivers));
+  RowVectorPtr result;
+  const uint64_t prevInjectedErrorCount = injectedErrorCount_;
+  try {
+    result = veloxToPrestoResult(execute(plan, maxDrivers));
+  } catch (VeloxRuntimeError& error) {
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:470]:
```
} catch (VeloxRuntimeError& error) {
  if (injectedErrorCount_ == prevInjectedErrorCount) {
    throw here
  }
  VELOX_CHECK_EQ
}
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 16]
```diff
@@ -292,6 +307,16 @@ void WriterFuzzer::go() {
   auto startTime = std::chrono::system_clock::now();
   size_t iteration = 0;
 
+  // Faulty fs will generate file system write error with certain possibility
+  if (FLAGS_file_system_error_injection) {
+    faultyFs_->setFileInjectionHook([&](FaultFileOperation* op) {
+      if (vectorFuzzer_.coinToss(1)) {
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:313]: 0.01? 1 means always injection

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 LGTM % minors. Thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 17]
```diff
@@ -436,7 +463,18 @@ void WriterFuzzer::verifyWriter(
...
-  const auto result = veloxToPrestoResult(execute(plan, maxDrivers));
+  RowVectorPtr result;
+  const uint64_t prevInjectedErrorCount = injectedErrorCount_;
+  try {
+    result = veloxToPrestoResult(execute(plan, maxDrivers));
+  } catch (VeloxRuntimeError& error) {
+    if (injectedErrorCount_ == prevInjectedErrorCount) {
+      throw error;
+    }
+    VELOX_CHECK(injectedErrorCount_ > prevInjectedErrorCount);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:474]:
```
VELOX_CHECK_GT(xxx, yyy, "Unexpected writer fuzzer failure: {}", error.message());
```

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 LGTM. Thanks for the update!

File: velox/dwio/common/tests/CMakeLists.txt [hunk 18]
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
 add_subdirectory(utils)
+velox_add_library(velox_dwio_faulty_file_sink FaultyFileSink.cpp)
+velox_link_libraries(velox_dwio_faulty_file_sink velox_file_test_utils
```
Comment (assignUser) [inline velox/dwio/common/tests/CMakeLists.txt:17]: This is a test utility right? In that case it should maybe live in `utils` but it should def use the base cmake functions `add_library` and `target_link_libraries`  to create a standalone lib, that way when we can install velox you can build with test stuff but the actual library is still 'clean'.

**Pull request closed without merging**